rules/onchain_budget.go

```diff
@@ -0,0 +1,694 @@
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	// Compile-time checks to ensure that OnChainBudget, OnChainBudgetMgr
+	// and OnChainBudgetEnforcer implement the appropriate Manager, Enforcer
+	// and Values interface.
+	_ Manager  = (*OnChainBudgetMgr)(nil)
+	_ Enforcer = (*OnChainBudgetEnforcer)(nil)
+	_ Values   = (*OnChainBudget)(nil)
+)
+
+const (
+	// OnChainBudgetName is the string identifier of the OnChainBudgetMgr
+	// rule.
+	OnChainBudgetName = "on-chain-budget"
+
+	// spendAmtKey is the key that will be used in the persisted KV store to
+	// store the total amount that has been spent.
+	onChainSpent = "spent-amt"
+
+	// pendingSpentKey is the key that will be used in the persisted KV
+	// store to keep track of the total pending spent amount.
+	onChainPending = "pending-spent"
+)
+
+// OnChainBudgetMgr manages the OnChainBudget rule.
+type OnChainBudgetMgr struct {
+	// The mutex is used to ensure that only one Enforcer created by the
+	// manager can run the HandleRequest and HandleResponse functions at
+	// any given time. This prevents db entry race conditions.
+	sync.Mutex
+}
+
+// Stop cleans up the resources held by the manager.
+//
+// NOTE: This is part of the Manager interface.
+func (o *OnChainBudgetMgr) Stop() error {
+	return nil
+}
+
+// NewEnforcer constructs a new OnChainBudgetEnforcer rule enforcer using the
+// passed values and config.
+//
+// NOTE: This is part of the Manager interface.
+func (o *OnChainBudgetMgr) NewEnforcer(cfg Config, values Values) (Enforcer,
+	error) {
+
+	budget, ok := values.(*OnChainBudget)
+	if !ok {
+		return nil, fmt.Errorf("values must be of type "+
+			"OnChainBudget, got %T", values)
+	}
+
+	return &OnChainBudgetEnforcer{
+		onChainBudgetConfig: cfg,
+		OnChainBudget:       budget,
+		OnChainBudgetMgr:    o,
+	}, nil
+}
+
+// NewValueFromProto converts the given proto value into a OnChainBudget Values
+// object.
+//
+// NOTE: This is part of the Manager interface.
+func (o *OnChainBudgetMgr) NewValueFromProto(v *litrpc.RuleValue) (Values,
+	error) {
+
+	rv, ok := v.Value.(*litrpc.RuleValue_OnChainBudget)
+	if !ok {
+		return nil, fmt.Errorf("incorrect RuleValue type %T", v.Value)
+	}
+
+	budget := rv.OnChainBudget
+
+	return &OnChainBudget{
+		AbsoluteAmtSats: budget.AbsoluteAmtSats,
+		MaxSatPerVByte:  budget.MaxSatPerVByte,
+	}, nil
+}
+
+// EmptyValue returns a new instance of OnChainBudget.
+//
+// NOTE: This is part of the Manager interface.
+func (o *OnChainBudgetMgr) EmptyValue() Values {
+	return &OnChainBudget{}
+}
+
+// onChainBudgetConfig is the config required by OnChainBudgetMgr. It can be
+// derived from the main rules Config struct.
+type onChainBudgetConfig interface {
+	GetStores() firewalldb.KVStores
+	GetReqID() int64
+	GetLndConnID() string
+}
+
+// OnChainBudgetEnforcer enforces requests and responses against a
+// OnChainBudget rule.
+type OnChainBudgetEnforcer struct {
```

Comment:
I want to check what our conclusion was regarding clearing all pending amounts and entries in the `onChainPending` local store, on restart of litd (i.e. consider them as confirmed even if they weren't)?

If I understood correctly, the conclusion was that it's nice to not clear them to have historical data showing where balance was deducted. Given that we also now add the memo field, that also gives some historical proof.

I agree with the decision to not clear them for now though, as when the memo field actually becomes useable, we can eventually clear the old pending amounts which were never succeeded in a channel opening, in the future.

Reply:
yeah. Worth noting that we cant really rely on the memo field not changing though. So this is a nice to have, temp measure. Worst case scenario - user just links a session and updates the budget